freshmaker/kojiservice.py

```diff
@@ -407,6 +407,41 @@ def get_bundle_csv(self, build_nvr):
         with zipfile.open(csv_files[0]) as f:
             return yaml.safe_load(f)
 
+    @region.cache_on_arguments()
+    def get_module_name_streams(self, build_nvr):
+        """
+        Get module's name and its streams through build NVR
+
+        :param str build_nvr: NVR of module build.
+        :return: a dict for module name and its streams, e.g. {"key": [values]}
+        :rtype: dict
+        """
+        module_name_streams_mapping = dict()
+        build = self.get_build(build_nvr)
+        modulemd_str = build.get("extra", {}).get("typeinfo", {}).get("module", {}).get("modulemd_str")
+        if modulemd_str:
+            try:
+                import gi
+                gi.require_version("Modulemd", "2.0")
+                from gi.repository import Modulemd
+                fedora_index = Modulemd.ModuleIndex.new()
+                ret, failures = fedora_index.update_from_string(modulemd_str, False)
+                if ret:
+                    merger = Modulemd.ModuleIndexMerger.new()
+                    merger.associate_index(fedora_index, 0)
+                    merged_index = merger.resolve()
+                    module_names = merged_index.get_module_names()
+                    for module_name in module_names:
+                        module = merged_index.get_module(module_name)
+                        streams = module.get_stream_names()
+                        module_name_streams_mapping[module_name] = streams
+                else:
+                    log.error("Update ModuleIndex failed from %s: %s", modulemd_str, str(failures))
+            except Exception as e:
+                log.error("Parse module name and streams failed from NVR %s: %s", build_nvr, str(e))
+
+        return module_name_streams_mapping
```

Comment:
When error happens, an empty dict is returned, it makes no sense, I'd prefer return `None` if we hit any issue of fetching the module metadata.

Reply:
I believe a module build can only have one stream, not 100% sure, but according to the API names, it should be true, you can get only one stream from a modulemd, there is a `get_stream_name` API, as you see, it's `name` not `names`.

I'd suggest refactoring this API to just return a `Modulemd.ModuleStream` instance, like:
```
def get_modulemd(self, build_nvr):
    ...
    try:
        mmd = Modulemd.ModuleStream.read_string(modulemd_str, False, None, None)
    ...
    # any error happens
    # return None
    return mmd
```
then you can get module name and stream with:
```
mmd = session.get_modulemd("foobar-flatpak-xxxxx")
module_name = mmd.get_module_name()
module_stream = mmd.get_stream_name()
```

Reply:
Thanks @qixiang for you carefully review and constructive suggestions. I check the doc again, and there is description "modulemd_module_stream_read_string has been deprecated since version 2.11 and should not be used in newly-written code.
Use modulemd_read_packager_string() instead.". I change it with read_packager_string, please have a look.